### PR TITLE
Put constant data in .rodata section instead of .text

### LINKED ISSUE #6029: put constant data in .rodata section instead of .text

### PR DESCRIPTION
**Pull request by Ersikan**
Closes #6029

This PR is still in progress. There are still no tests and a few unresolved issues.

## Unresolved issues

Is this a good design ?

Should we drop the `Text` part of `TextBlock` and `TextBlockList` as it isn't just about the `.text` section anymore ?

Should the program headers for GOT and `.rodata` be merged ? They have strictly the same configuration, but maybe it would hurt incremental compilation as we would need to move them around more often when they grow.

### CODE REVIEW

File: src/link/Elf.zig [hunk 1]
```diff
@@ -489,6 +507,29 @@ pub fn populateMissingMetadata(self: *Elf) !void {
...
     }
+    if (self.phdr_load_ro_index == null) {
+        self.phdr_load_ro_index = @intCast(u16, self.program_headers.items.len);
+        self.rodata_block_list.phdr_index = self.phdr_load_ro_index;
+        // TODO Find a hint about how much data need to be in rodata ?
+        const file_size = 1024;
+        // Same reason as for GOT
+        const p_align = if (self.base.options.target.os.tag == .linux) 0x1000 else @as(u16, ptr_size);
+        const off = self.findFreeSpace(file_size, p_align);
+        log.debug("found PT_LOAD free space 0x{x} to 0x{x}\n", .{ off, off + file_size });
```
Comment (g-w1) [inline src/link/Elf.zig:518]: Log automatically inserts a "\n" so you don't need another.

Comment (FireFox317) [reply src/link/Elf.zig:518]: All the other `log.debug` lines in that file also have the "\n", so "\n" here is fine too.

File: src/link/Elf.zig [hunk 2]
```diff
@@ -2105,11 +2164,22 @@ fn allocateTextBlock(self: *Elf, text_block: *TextBlock, new_block_size: u64, al
         text_block.next = null;
     }
     if (free_list_removal) |i| {
-        _ = self.text_block_free_list.swapRemove(i);
+        _ = block_list.free_list.swapRemove(i);
     }
     return vaddr;
 }
 
+/// Get the block list corresponding to a specifi decl
```
Comment (g-w1) [inline src/link/Elf.zig:2172]: Nit: typo

File: src/link/Elf.zig (hunk 1 above)
Comment (g-w1) [reply src/link/Elf.zig:518]: It is a mistake https://discord.com/channels/605571803288698900/785499283368706060/818987635883114547 
```
[6:24 PM]
andrewrk:
the \n is a mistake in all the logs
[6:24 PM]
andrewrk:
if you copied it from me then you copied my typo :slight_smile:
```

Comment (Ersikan) [reply src/link/Elf.zig]: I removed the "\n" from all the logs.

File: src/link/Elf.zig (hunk 2 above)
Comment (Ersikan) [reply src/link/Elf.zig]: Fixed

File: src/link/Elf.zig [hunk 3]
```diff
@@ -2221,7 +2228,7 @@ pub fn allocateDeclIndexes(self: *Elf, decl: *Module.Decl) !void {
 pub fn freeDecl(self: *Elf, decl: *Module.Decl) void {
     if (self.llvm_ir_module) |_| return;
 
-    const block_list = self.getDeclBlockList(decl);
+    const block_list = self.getDeclBlockList(decl) orelse return;
```
Comment (Ersikan) [inline src/link/Elf.zig:2231]: I'm really not sure that here we should return from the function if the Decl is still not typed, as before this commit the function `freeDecl` did not check it.

Is it guaranteed that a Decl has an associated TextBlock only if it has already been typed ?

File: src/link/Elf.zig [hunk 4]
```diff
@@ -201,6 +202,21 @@ pub const TextBlock = struct {
...
+    last_block: ?*TextBlock,
+    phdr_index: ?u16,
+    section_index: ?u16,
+
+    const empty = TextBlockList{
+        .free_list = .{},
+        .last_block = null,
+        .phdr_index = null,
+        .section_index = null,
+    };
```
Comment (Vexu) [inline src/link/Elf.zig:217]:
```suggestion
    free_list: std.ArrayListUnmanaged(*TextBlock) = .{},
    last_block: ?*TextBlock = null,
    phdr_index: ?u16 = null,
    section_index: ?u16 = null,
```
We have default values for this purpose, `TextBlockList.empty` can be replaced with `.{}`.

Comment (Ersikan) [reply src/link/Elf.zig]: Thank you for the review ! I think I did this because the `TextBlock` struct does the same, and just did not think it through.

**Comment by andrewrk** [comment]
30+ days old draft; closing. are you sure this wasn't ready for review?

**Comment by Ersikan** [comment]
@andrewrk Not long after starting this PR I started an internship and I did not have time to finish working on it.

When I have more time, I'll continue working on it. However, in the meantime this PR may be closed. I will open a new one when the time comes.

**Comment by andrewrk** [comment]
Good luck with the internship!

**Comment by ifreund** [comment]
@Ersikan You mentioned that you intend to continue working on this on discord but didn't have a "reopen" button. I've went ahead and solved that problem for you I hope :)

**Comment by andrewrk** [comment]
30+ days old draft; closing. Please open a new pull request when this is *ready for review*.

**Pull request closed without merging**